### PR TITLE
Container State Machine

### PR DESCRIPTION
**Pull request by kflansburg**

### CODE REVIEW

**Comment by kflansburg** [comment]
Not quite ready to merge but I think it’s a good time for some reviews. What is working:

* Refactored state API to be generic.
* Rebased with master.
* Implemented WASI container state machine. This was nasty but I really like how simple the state handlers are now.

Still left to do:
* wascc container state machine.
* Some todos that I identified which probably don’t need to go into this PR.
* Remove some unwraps

**Review by thomastaylor312 — COMMENTED** [review]
This is looking fairly solid, think we should probably get @itowlson's review on this too since he has been mucking around in here recently. I actually like the combined shared state thing as well

File: crates/kubelet/src/container/state.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+pub mod prelude {
+    pub use crate::container::{Container, Handle, Status};
+    pub use crate::state::{ResourceState, SharedState, State, Transition, TransitionTo};
+}
+
+/// Iteratively evaluate state machine until it returns Complete.
+pub async fn run_to_completion<S: ResourceState<Manifest = Container, Status = Status>>(
+    client: &kube::Client,
+    initial_state: impl State<S>,
+    shared_state: SharedState<S::SharedState>,
```
[the remark this answers is not in the document]
Comment (kflansburg) [reply crates/kubelet/src/container/state.rs:21]: It is definitely confusing. I wonder if the `State` trait should actually be like `Handler`, and we might be able to just called it `shared` or something. Any ideas?

File: crates/kubelet/src/lib.rs [hunk 2]
```diff
@@ -10,7 +10,11 @@
 //! use kubelet::config::Config;
 //! use kubelet::pod::Pod;
 //! use kubelet::provider::Provider;
-//! use kubelet::state::{SharedState, Stub, AsyncDrop};
+//! use kubelet::state::{SharedState};
```
[the remark this answers is not in the document]
Comment (kflansburg) [reply crates/kubelet/src/lib.rs:13]: I was going to agree, but on second thought I think we should stick to what `rustfmt` enforces. Adding additional style opinions would be a lot of manual labor to enforce consistently. Is that ok?

File: crates/kubelet/src/pod/status.rs [hunk 3]
```diff
@@ -153,46 +142,83 @@ pub fn make_status_with_containers(
...
+    /// Create a new status with no fields set.
+    pub fn new() -> Self {
+        Status(Default::default())
+    }
 
-impl Default for StatusMessage {
-    fn default() -> Self {
-        Self::LeaveUnchanged
+    /// Set Pod phase.
+    pub fn set_phase(&mut self, phase: Phase) {
```
[the remark this answers is not in the document]
Comment (kflansburg) [reply crates/kubelet/src/pod/status.rs:165]: I think we could basically rename this struct to `StatusBuilder` then. I can if you want. The reason I did this is that there is a distinction between a field that was never set, or is explicitly being set to "", this results in patches overwriting things they shouldnt. There might be a simpler way to do this, and I didnt do the same thing for containers because its much less important. Thoughts?

File: crates/kubelet/src/pod/status.rs [hunk 4]
```diff
@@ -153,46 +142,83 @@ pub fn make_status_with_containers(
...
+    /// Set Pod init container statuses.
+    pub fn set_init_container_statuses(
+        &mut self,
+        init_container_statuses: Vec<KubeContainerStatus>,
+    ) {
+        self.0.init_container_statuses = Some(init_container_statuses);
+    }
+
+    pub(crate) fn into_json(self) -> serde_json::Value {
+        let mut status = serde_json::Map::new();
```
[the remark this answers is not in the document]
Comment (kflansburg) [reply crates/kubelet/src/pod/status.rs:193]: See above. Calling `json!` doesnt really create a patch. It may be that what we really want is to use `jsondiff` here, but then you need to have an updated version of the Pod status and then there are race conditions, and this actually seems simpler.

File: crates/kubelet/src/pod/status.rs [hunk 5]
```diff
@@ -213,6 +239,12 @@ pub enum Phase {
     Unknown,
 }
 
+impl std::fmt::Display for Phase {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", serde_json::json!(self).as_str().unwrap())
```
[the remark this answers is not in the document]
Comment (kflansburg) [reply crates/kubelet/src/pod/status.rs:244]: I'm not sure what implementation you are referring to. I don't see that `PodStatus` implements `Display`, and phase is simply stored as a `String`. The reason I did it this way is that as far as I know this is the only way to automatically generate `Display` for `enums`, rather than having to specify the exact string used. You can also use `Debug`, but `serde` lets us control the capitalization, etc. a little more. I don't think this can panic. I can change this to use `Debug` but I don't like the idea of having to hard code the strings and update them if the `enum` changes.

File: crates/kubelet/src/state/common/mod.rs [hunk 6]
```diff
@@ -50,21 +52,21 @@ pub trait GenericProviderState: 'static + Send + Sync {
 /// Exposes pod state in a way that can be consumed by
 /// the generic states.
 #[async_trait::async_trait]
-pub trait GenericPodState: 'static + Send + Sync {
+pub trait GenericPodState: ResourceState<Manifest = Pod, Status = PodStatus> {
     /// Stores the pod module binaries for future execution. Typically your
     /// implementation can just move the modules map into a member field.
-    fn set_modules(&mut self, modules: HashMap<String, Vec<u8>>);
+    async fn set_modules(&mut self, modules: HashMap<String, Vec<u8>>);
```
[the remark this answers is not in the document]
Comment (kflansburg) [reply crates/kubelet/src/state/common/mod.rs:58]: So, I ran into this because the `ModuleRunContext` had to be moved into a `SharedState` so that individual container state machines could load their module data. I'm glad you brought this up because I think that we should make pretty much all public trait methods `async`, the reason being that we have a lot of `async` methods and Tokio `RwLocks` floating around and it is really annoying as a provider author to realize that you can't use something `async` in an `async` trait just because we (the maintainers) didn't forsee that use case. Thoughts?

File: crates/kubelet/src/container/state.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (thomastaylor312) [reply crates/kubelet/src/container/state.rs:21]: Because this is billed as a "state machine" (which it is), I think the name "state" should remain with the actual state type. However, just using `shared` might be just fine

File: crates/kubelet/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (thomastaylor312) [reply crates/kubelet/src/lib.rs:13]: I don't think `rustfmt` enforces this. I very rarely see single imports using braces and rustfmt doesn't correct me if I don't have them

File: crates/kubelet/src/pod/status.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (thomastaylor312) [reply crates/kubelet/src/pod/status.rs:165]: I personally think this:

```rust
let status = StatusBuilder::new().phase(...).message(...).build();
```

reads much better (and is equivalent to):

```rust
let mut status = Status::new();
status.set_phase(...);
status.set_message(...);
```

I figure the point about an empty field is pretty easy to handle. You can still use the setter/builder to explicitly set an empty string and leave it empty otherwise. However, this is completely a bikeshedding, so feel no pressure to change it

File: crates/kubelet/src/pod/status.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (thomastaylor312) [reply crates/kubelet/src/pod/status.rs:193]: Ah, ok valid point

File: crates/kubelet/src/pod/status.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (thomastaylor312) [reply crates/kubelet/src/pod/status.rs:244]: Oh wait, I somehow misread `Phase` as `Status` when I was going through your code 🤦. Sorry

File: crates/kubelet/src/state/common/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (thomastaylor312) [reply crates/kubelet/src/state/common/mod.rs:58]: I figured that may be part of the reasoning behind this and I am fine with it. But I also try to avoid having things be async if they don't have to be. The safe bet is to probably keep it async and then we can pull that back before 1.0 if it truly isn't being used as async

File: crates/kubelet/src/container/state.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (kflansburg) [reply crates/kubelet/src/container/state.rs:21]: Will do.

File: crates/kubelet/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kflansburg) [reply crates/kubelet/src/lib.rs:13]: Yeah, I think my point is that we cant automatically enforce that style either way. I will remove it here but I usually think that opinions on style held by a project should be something that can be automatically linted.

File: crates/kubelet/src/pod/status.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kflansburg) [reply crates/kubelet/src/pod/status.rs:165]: Yep, I will refactor it to a Builder, I was just explaining why it got so complicated.

File: crates/kubelet/src/state/common/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (kflansburg) [reply crates/kubelet/src/state/common/mod.rs:58]: Sounds good.

File: crates/kubelet/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kflansburg) [reply crates/kubelet/src/lib.rs:13]: For posterity, I used this to find one other occurrence:
```
grep -rE . -e '{([a-zA-Z]+)}'
```
Something like that could be added to our test suite.

**Comment by kflansburg** [comment]
@thomastaylor312 @itowlson  I've added the wascc provider and cleaned up the outstanding panics to push this across the finish line. I think there are some opportunities to improve here by:

1. Refactoring some of the larger state handlers into multiple methods / macros for error transitions.
1. Pod::Running and container::Terminated seem ripe for a generic implementation. 

I would prefer to merge for now though and open tickets to keep this contained (no pun intended). I would prefer to merge this before starting any work on pulling out an operator framework.

File: crates/wascc-provider/src/states/container/waiting.rs [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+                                state.pod.name(),
+                                container.name(),
+                                e
+                            ),
+                            true,
+                        ),
+                    )
+                }
+            }
+        };
```
Comment (itowlson) [inline crates/wascc-provider/src/states/container/waiting.rs:124]: ~I find this much nesting really hard to parse.  Can we simplify this in any way?~  Looking at the broader context I now suspect this is existing code and we are just moving it into a state with the minimum possible change and risk, so can probably ignore this for now.

File: crates/wascc-provider/src/states/container/waiting.rs [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+                            e
+                        ),
+                        true,
+                    ),
+                )
+            }
+        }
+
+        Transition::next(self, Running)
+    }
```
Comment (itowlson) [inline crates/wascc-provider/src/states/container/waiting.rs:263]: Unless I've missed something, this is a 160-line method with up to 7 levels of indentation in places. Maybe I am dumb but I find that really hard to keep straight in my mind.  I assume that this is probably code that already exists and we are just moving it into a state, so I can see that we might not want to change it as part of this PR, but it should be high on some list for refactoring!

File: crates/kubelet/src/state/common/mod.rs (hunk 6 above)
Comment (itowlson) [inline crates/kubelet/src/state/common/mod.rs:58]: Making these async feels like it should be a separate design discussion and a separate commit, not rolled into the container state changes.

File: crates/kubelet/src/state/common/mod.rs [hunk 9]
```diff
@@ -73,11 +75,11 @@ pub trait GenericProvider: 'static + Send + Sync {
     /// The state of the provider itself.
     type ProviderState: GenericProviderState;
     /// The state that is passed between Pod state handlers.
-    type PodState: GenericPodState;
+    type PodState: GenericPodState + ResourceState<SharedState = Self::ProviderState>;
```
Comment (itowlson) [inline crates/kubelet/src/state/common/mod.rs:78]: Again I kind of feel like this is changing the design of one component in a PR that is concerned with another component.  But I guess that it emerges from a need to model shared state for containers.

File: crates/kubelet/src/kubelet.rs [hunk 10]
```diff
@@ -255,16 +256,18 @@ mod test {
     struct PodState;
 
     #[async_trait::async_trait]
-    impl AsyncDrop for PodState {
-        type ProviderState = ProviderState;
+    impl ResourceState for PodState {
+        type Manifest = Pod;
+        type Status = Status;
+        type SharedState = ProviderState;
```
Comment (itowlson) [inline crates/kubelet/src/kubelet.rs:262]: This is an elegant solution to the `async_drop` problem and I like it **much** better than mine - thank you!

**Review by itowlson — APPROVED** [review]
This is great to see.  Thanks!

**Review by thomastaylor312 — APPROVED** [review]

File: crates/wascc-provider/src/states/container/waiting.rs (hunk 7 above)
Comment (kflansburg) [reply crates/wascc-provider/src/states/container/waiting.rs:124]: I definitely think these things should be broken up in the future.

File: crates/wascc-provider/src/states/container/waiting.rs (hunk 8 above)
Comment (kflansburg) [reply crates/wascc-provider/src/states/container/waiting.rs:263]: Yep, I essentially just pulled in `start_container` and included any necessary context from `Pod::starting`. I think these should be broken up in future refactoring.

File: crates/kubelet/src/state/common/mod.rs (hunk 6 above)
Comment (kflansburg) [reply crates/kubelet/src/state/common/mod.rs:58]: I think the reason this triggered the discussion is that the container state machines need access to these data structures concurrently. I think there is an argument to be made for moving the image pull into the container state machine so that they do not need this shared reference. 

I think we should establish a policy of all-async, but we can discuss that later.

File: crates/kubelet/src/state/common/mod.rs (hunk 9 above)
Comment (kflansburg) [reply crates/kubelet/src/state/common/mod.rs:78]: This was a little weird because the shared state type does need to be specified in `ResourceState`, but it was also needed on `Provider` for ergonomics. This line was changed to enforce that both of those types were the same.

**Comment by kflansburg** [comment]
Thanks for reviewing @itowlson @thomastaylor312 !
